File: LeetCode/1038-binary-search-tree-to-greater-sum-tree/1038-binary-search-tree-to-greater-sum-tree.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    void revinorder(TreeNode *root, int &sum) {
        if (!root) {
            return;
        }
        revinorder(root->right, sum);
        sum += root->val;
        root->val = sum;
        revinorder(root->left, sum);
    }
    TreeNode* bstToGst(TreeNode* root) {
        int sum = 0;
        revinorder(root, sum);
        return root;
    }
};





// /**
//  * Definition for a binary tree node.
//  * struct TreeNode {
//  *     int val;
//  *     TreeNode *left;
//  *     TreeNode *right;
//  *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
//  *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
//  *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
//  * };
//  */
// class Solution {
// public:
//     vector<int> temp;
//     void inorder(TreeNode* root) {
//         if (!root) {
//             return;
//         }
//         inorder(root->left);
//         temp.push_back(root->val);
//         inorder(root->right);
//     }
//     int bs(int val, vector<int> &ans) {
//         int low = 0, high = ans.size() - 1;
//         while (low <= high) {
//             int mid = low + (high - low) / 2;
//             if (temp[mid] == val) {
//                 return ans[mid];
//             } else if (temp[mid] > val) {
//                 high = mid - 1;
//             } else {
//                 low = mid + 1;
//             }
//         }
//         return val;
//     }
//     void solve(TreeNode* root, vector<int> &ans) {
//         if (!root) {
//             return;
//         }
//         root->val = bs(root->val, ans);
//         solve(root->left, ans);
//         solve(root->right, ans);
//     }
//     TreeNode* bstToGst(TreeNode* root) {
//         inorder(root);
//         int n = temp.size();
//         vector<int> ans(n, 0);
//         ans[n-1] = temp[n-1];
//         for (int i = n-2; i >= 0; i--) {
//             ans[i] = temp[i] + ans[i+1];
//         }
//         solve(root, ans);
//         return root;
//     }
// };

File: LeetCode/1038-binary-search-tree-to-greater-sum-tree/1038-binary-search-tree-to-greater-sum-tree.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def __init__(self):
        self.sum = 0

    def revinorder(self, root: TreeNode):
        if not root:
            return
        self.revinorder(root.right)
        self.sum += root.val
        root.val = self.sum
        self.revinorder(root.left)

    def bstToGst(self, root: TreeNode) -> TreeNode:
        self.sum = 0
        self.revinorder(root)
        return root




# # Definition for a binary tree node.
# # class TreeNode:
# #     def __init__(self, val=0, left=None, right=None):
# #         self.val = val
# #         self.left = left
# #         self.right = right
# class Solution:
#     def __init__(self):
#         self.temp = []
    
#     def inorder(self, root: TreeNode):
#         if root is None:
#             return
#         self.inorder(root.left)
#         self.temp.append(root.val)
#         self.inorder(root.right)
    
#     def bs(self, val: int, ans: list[int]) -> int:
#         low, high = 0, len(ans) - 1
#         while low <= high:
#             mid = (low + high) // 2
#             if self.temp[mid] == val:
#                 return ans[mid]
#             elif self.temp[mid] > val:
#                 high = mid - 1
#             else:
#                 low = mid + 1
#         return val
    
#     def solve(self, root: TreeNode, ans: list[int]):
#         if root is None:
#             return
#         root.val = self.bs(root.val, ans)
#         self.solve(root.left, ans)
#         self.solve(root.right, ans)
    
#     def bstToGst(self, root: TreeNode) -> TreeNode:
#         self.inorder(root)
#         n = len(self.temp)
#         ans = [0] * n
#         ans[n - 1] = self.temp[n - 1]
#         for i in range(n - 2, -1, -1):
#             ans[i] = self.temp[i] + ans[i + 1]
#         self.solve(root, ans)
#         return root